### ISSUE TITLE
bind-value-oninput throws

### ISSUE THREAD

**Issue by Andrzej-W** [aspnet/AspNetCore#5521]
Currently when we use `<input bind="@myVariable" />` variable is updated when input field looses focus (onchange event is called). Popular requirement in many applications is to implement incremental search. To do this we need:

- binding which updates variable after every input field change
- throttling - this is necessary if we want to do incremental search on the server or we have a big collection to filter; we should be able to specify a delay in milliseconds and our variable should be updated only after user stops typing for the specified period of time

I believe that throttling can be useful for other events also, for example mouse move, mouse wheel scrolling, etc.

**Comment by Andrzej-W** [comment]
I have just retested this bug in Blazor 0.7 and this code now works as expected:
```html
<input bind-value-oninput="@FieldValue" />
<p>@FieldValue</p>
@functions{
    string FieldValue { get; set; }
}
```
Unfortunately `<input bind-value-onkeydown="@FieldValue" />` and `<input bind-value-onkeyup="@FieldValue" />` throws `Error: System.InvalidCastException: Specified cast is not valid.`

**Comment by springy76** [comment]
Where can I read (in the docs) about `bind-x-y` using 2 hyphens and what it actually does?

https://blazor.net/docs/components/index.html#data-binding only knows about `bind-{property}`.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us.
@Andrzej-W, is there still a problem here or should this be closed now?

**Comment by Andrzej-W** [comment]
@mkArtakMSFT The problem with bind-value-oninput was fixed but if I want to use other events, for example bind-value-onkeydown, bind-value-onkeyup I'm still getting an exception. Currently I have the newest ASP.NET Core 3.0 Preview 4 and I see this in the browser's console:

```
WASM: ﻿Unhandled exception rendering component: blazor.webassembly.js:1:32082
WASM: System.ArgumentException: Object of type 'Microsoft.AspNetCore.Components.UIKeyboardEventArgs' cannot be converted to type 'Microsoft.AspNetCore.Components.UIChangeEventArgs'. blazor.webassembly.js:1:32082
WASM:   at System.RuntimeType.CheckValue (System.Object value, System.Reflection.Binder binder, System.Globalization.CultureInfo culture, System.Reflection.BindingFlags invokeAttr) <0x1a505a0 + 0x000ca> in <46556de4d85b4f28892019a3d1d69868>:0 blazor.webassembly.js:1:32082
WASM:   at System.Reflection.RuntimeMethodInfo.ConvertValues (System.Reflection.Binder binder, System.Object[] args, System.Reflection.ParameterInfo[] pinfo, System.Globalization.CultureInfo culture, System.Reflection.BindingFlags invokeAttr) <0x1a501d0 + 0x000f2> in <46556de4d85b4f28892019a3d1d69868>:0 blazor.webassembly.js:1:32082
WASM:   at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) <0x1a4f698 + 0x00088> in <46556de4d85b4f28892019a3d1d69868>:0 blazor.webassembly.js:1:32082
WASM:   at System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) <0x1a4f128 + 0x00016> in <46556de4d85b4f28892019a3d1d69868>:0 blazor.webassembly.js:1:32082
WASM:   at System.Delegate.DynamicInvokeImpl (System.Object[] args) <0x200a348 + 0x0019e> in <46556de4d85b4f28892019a3d1d69868>:0 blazor.webassembly.js:1:32082
WASM:   at System.MulticastDelegate.DynamicInvokeImpl (System.Object[] args) <0x200a0e0 + 0x00016> in <46556de4d85b4f28892019a3d1d69868>:0 blazor.webassembly.js:1:32082
WASM:   at System.Delegate.DynamicInvoke (System.Object[] args) <0x2009fe8 + 0x0000c> in <46556de4d85b4f28892019a3d1d69868>:0 blazor.webassembly.js:1:32082
WASM:   at Microsoft.AspNetCore.Components.EventCallbackWorkItem.InvokeAsync[T] (System.MulticastDelegate delegate, T arg) <0x2009dc0 + 0x000cc> in <b90d61056671486d8836d104b91086c9>:0 blazor.webassembly.js:1:32082
WASM:   at Microsoft.AspNetCore.Components.EventCallbackWorkItem.InvokeAsync (System.Object arg) <0x2009a90 + 0x0000c> in <b90d61056671486d8836d104b91086c9>:0 blazor.webassembly.js:1:32082
WASM:   at Microsoft.AspNetCore.Components.ComponentBase.Microsoft.AspNetCore.Components.IHandleEvent.HandleEventAsync (Microsoft.AspNetCore.Components.EventCallbackWorkItem callback, System.Object arg) <0x20099f8 + 0x0000a> in <b90d61056671486d8836d104b91086c9>:0 blazor.webassembly.js:1:32082
WASM:   at Microsoft.AspNetCore.Components.EventCallback.InvokeAsync (System.Object arg) <0x20095e8 + 0x00046> in <b90d61056671486d8836d104b91086c9>:0 blazor.webassembly.js:1:32082
WASM:   at Microsoft.AspNetCore.Components.Rendering.Renderer.DispatchEventAsync (System.Int32 eventHandlerId, Microsoft.AspNetCore.Components.UIEventArgs eventArgs) <0x2008f08 + 0x0005a> in <b90d61056671486d8836d104b91086c9>:0 blazor.webassembly.js:1:32082
Error: System.NullReferenceException: Object reference not set to an instance of an object.
  at Microsoft.AspNetCore.Components.Rendering.Renderer.GetErrorHandledTask (System.Threading.Tasks.Task taskToHandle) <0x2043c48 + 0x00112> in <b90d61056671486d8836d104b91086c9>:0 blazor.webassembly.js:1:6160
```

**Comment by danroth27** [comment]
@Andrzej-W `onkeydown` and `onkeyup` don't actually provide the value, so the behavior described is expected. @guardrex We should document this better. We may also look at improving this experience as we look at adding support for directive attributes.

Closing as we don't think there is anything left to do here.

**Comment by PirasannaRavi** [comment]
[https://github.com/aspnet/Blazor/issues/1235#issuecomment-489328205](url)

`bind-value-oninput` or `bind` does not work with `contenteditable` divs.

In the below code, the expectation is as and when I edit the div content I need to update the model so that I can then later save it back.
`<div contenteditable="true" bind-value-oninput="@contentEditableText">
                                               You can Edit this Text
                                            </div>
                                            @contentEditableText`

`@functions { string contentEditableText = "You can Edit this Text"; }`

**Comment by guardrex** [comment]
Hello @PirasannaRavi ... Please open product feedback over in the [aspnet/AspNetCore issues](https://github.com/aspnet/AspNetCore/issues).

**Comment by PirasannaRavi** [comment]
Sure @guardrex.  
I created a new Issue as suggested. 
[https://github.com/aspnet/AspNetCore/issues/9974](url)

**Comment by mvr-devs** [comment]
bind-value-oninput does not throw an exception but it also does not update the variable as of version 3.0.0-preview6.19307.2

```
<div class="search-bar">
    <input bind-value-oninput="@SearchTerm" oninput="@(()=>PreformSearch())" placeholder="search for customers or tickets" />.
</div>
```


```
@code {

    private string SearchTerm { get; set; }

    private void PreformSearch()
    {
        Console.WriteLine($"Something tells me you are trying to search for '{SearchTerm}' jackleg-joe...");
    }

}
```

**Comment by guardrex** [comment]
@mvr-devs Try this ...

```cshtml
<div class="search-bar">
    <input @oninput="@PreformSearch" placeholder="search for customers or tickets" />
</div>

@code {
    private void PreformSearch(UIChangeEventArgs e)
    {
        Console.WriteLine($"Something tells me you are trying to search for '{e.Value}' ...");
    }
}
```

**Comment by mvr-devs** [comment]
Perfect !!  Thanks so much !!
